Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -358,91 +377,129 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
             }
             return 0
         }
     }
 
-    public func decoderRemoved(ctx: ChannelHandlerContext) {
-        // Remove the stored reference to ChannelHandlerContext
-        parser.data = UnsafeMutableRawPointer(bitPattern: 0x0000deadbeef0000)
-
-        // Set the callbacks to nil as we dont need these anymore
-        settings.on_body = nil
-        settings.on_chunk_complete = nil
-        settings.on_url = nil
-        settings.on_status = nil
-        settings.on_chunk_header = nil
-        settings.on_chunk_complete = nil
-        settings.on_header_field = nil
-        settings.on_header_value = nil
-        settings.on_message_begin = nil
-    }
+    // Decode HTTP until there is nothing more to decode.
+    private func decodeHTTP(ctx: ChannelHandlerContext) throws {
+        // We need to refetch the cumulationBuffer on each loop as it may has changed due re-entrance calls of channelRead(...)
+        while let bufferSlice = self.cumulationBuffer, bufferSlice.readableBytes > 0 {
+            let result = try bufferSlice.withVeryUnsafeBytes { (pointer) -> size_t in
+                state.baseAddress = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
 
-    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
-        if let slice = state.slice {
-            // If we stored a slice before we need to ensure we move the readerIndex so we don't try to parse the data again. We
-            // also need to update the reader index to whatever it is now.
-            state.slice = (buffer.readerIndex, slice.length)
-            buffer.moveReaderIndex(forwardBy: state.readerIndexAdjustment)
-        }
+                let result = state.baseAddress!.withMemoryRebound(to: Int8.self, capacity: pointer.count, { p in
+                    c_nio_http_parser_execute(&parser, &settings, p.advanced(by: bufferSlice.readerIndex), bufferSlice.readableBytes)
+                })
 
-        let result = try buffer.withVeryUnsafeBytes { (pointer) -> size_t in
-            state.baseAddress = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+                state.baseAddress = nil
 
-            let result = state.baseAddress!.withMemoryRebound(to: Int8.self, capacity: pointer.count, { p in
-                c_nio_http_parser_execute(&parser, &settings, p.advanced(by: buffer.readerIndex), buffer.readableBytes)
-            })
+                if let error = state.currentError {
+                    throw error
+                }
 
-            if let error = state.currentError {
-                throw error
+                let httpError = parser.http_errno
+                if httpError != 0 {
+                    self.state.currentError = HTTPParserError.httpError(fromCHTTPParserErrno: http_errno(rawValue: httpError))!
+                    throw self.state.currentError!
+                }
+                return result
             }
```

Comment:
This block seems wildly too long to me: do we need the pointer to be valid for this long?

Reply:
nope let me move some stuff outside.